src/main/java/net/glowstone/block/itemtype/ItemBow.java

```diff
@@ -0,0 +1,115 @@
+package net.glowstone.block.itemtype;
+
+import java.util.concurrent.ThreadLocalRandom;
+import net.glowstone.entity.GlowPlayer;
+import net.glowstone.util.InventoryUtil;
+import org.bukkit.GameMode;
+import org.bukkit.Material;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.entity.Arrow;
+import org.bukkit.entity.SpectralArrow;
+import org.bukkit.entity.TippedArrow;
+import org.bukkit.inventory.ItemStack;
+
+public class ItemBow extends ItemTimedUsage {
+    private static final long TICKS_TO_FULLY_CHARGE = 20;
+    private static final double MAX_BASE_DAMAGE = 9;
+    private static final double MAX_SPEED = 53;
+
+    @Override
+    public void startUse(GlowPlayer player, ItemStack item) {
+        boolean haveArrow = false;
+        findArrow: for (ItemStack itemStack : player.getInventory().getContents()) {
+            switch (itemStack.getType()) {
+                case ARROW:
+                case SPECTRAL_ARROW:
+                case TIPPED_ARROW:
+                    haveArrow = true;
+                    break findArrow;
+                default:
+                    // do nothing
+            }
+        }
+        if (haveArrow) {
+            player.setUsageItem(item);
+            player.setUsageTime(TICKS_TO_FULLY_CHARGE);
+        }
+    }
+
+    @Override
+    public void endUse(GlowPlayer player, ItemStack item) {
+        ItemStack currentArrow = null;
+        // Check arrows again, since plugins can change the inventory while a bow is drawn
+        findArrow: for (ItemStack itemStack : player.getInventory().getContents()) {
```

Comment:
Merge both for-loops into a method, returning an `Optional<ItemStack>`.

Reply:
Done in 4a56ae374ea310c61123396fc796d71a0692da34